torchrl/objectives/dreamer.py

```diff
@@ -111,11 +140,59 @@ def __init__(
         self.free_nats = free_nats
         self.delayed_clamp = delayed_clamp
         self.global_average = global_average
+        # actor
+        self.actor_model = actor_model
+        self.value_model = value_model
+        self.model_based_env = model_based_env
+        self.imagination_horizon = imagination_horizon
+        self.discount_loss = discount_loss
+        if gamma is not None:
+            warnings.warn(_GAMMA_LMBDA_DEPREC_WARNING, category=DeprecationWarning)
+            self.gamma = gamma
+        if lmbda is not None:
+            warnings.warn(_GAMMA_LMBDA_DEPREC_WARNING, category=DeprecationWarning)
+            self.lmbda = lmbda
+        # value
+        self.value_loss_function = value_loss if value_loss is not None else "l2"
+        self.value_gamma = value_gamma
+        self.value_discount_loss = value_discount_loss
 
     def _forward_value_estimator_keys(self, **kwargs) -> None:
-        pass
+        if self._value_estimator is not None:
+            self._value_estimator.set_keys(
+                value=self._tensor_keys.value,
+            )
 
     def forward(self, tensordict: TensorDict) -> torch.Tensor:
+        world_tensordict, sampled_tensordict_world = self.model_loss(tensordict)
+        with autocast(dtype=torch.float16):
+            actor_tensordict, sampled_tensordict_actor = self.actor_loss(
+                sampled_tensordict_world
+            )
+            value_tensordict, sampled_tensordict_value = self.value_loss(
+                sampled_tensordict_actor
+            )
+        # result_tensordict.update(actor_tensordict) # Ultimately we could all lossed together everything,
+        # result_tensordict.update(value_tensordict) # but for the sake of simplicity of demonstration I pass the result as raw as possible
+        return (
+            world_tensordict,
+            actor_tensordict,
+            value_tensordict,
+            sampled_tensordict_world,
+            sampled_tensordict_actor,
+            sampled_tensordict_value,
+        )
```

Comment:
these should be nested in a single tensordict.
We usually put things together like this:
```python
TensorDict({
   "loss_actor": val,
   "loss_value": other_val,
    ....}, [])
```